File: implementations/java/server/src/main/java/com/theshuai/tunnelserver/management/service/UserDiagramDocumentService.java
package com.theshuai.tunnelserver.management.service;

import com.theshuai.tunnelserver.management.model.UserDiagramDocument;
import com.theshuai.tunnelserver.management.repository.UserDiagramDocumentRepository;
import com.theshuai.tunnelserver.management.repository.UserDiagramDocumentRepository.UserDiagramDocumentSummary;
import com.theshuai.tunnelserver.management.security.ManagementContext;
import org.springframework.http.HttpStatus;
import org.springframework.orm.ObjectOptimisticLockingFailureException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;
import org.springframework.web.server.ResponseStatusException;

import java.time.Instant;
import java.util.Base64;
import java.util.List;

@Service
public class UserDiagramDocumentService {
    private static final int MAX_DOCUMENTS_PER_USER = 100;
    private static final int MAX_SNAPSHOT_BYTES = 3 * 1024 * 1024;

    private final UserDiagramDocumentRepository repository;

    public UserDiagramDocumentService(UserDiagramDocumentRepository repository) {
        this.repository = repository;
    }

    @Transactional(readOnly = true)
    public List<DiagramDocumentView> list(ManagementContext context) {
        Owner owner = owner(context);
        return repository.findSummariesByOwner(owner.tenantId(), owner.username()).stream()
                .map(this::view)
                .toList();
    }

    @Transactional(readOnly = true)
    public DiagramDocumentDetail get(ManagementContext context, long id) {
        UserDiagramDocument document = requireOwned(context, id);
        return new DiagramDocumentDetail(view(document), Base64.getEncoder().encodeToString(document.getSnapshotData()));
    }

    @Transactional
    public DiagramDocumentView create(ManagementContext context, DiagramDocumentMutation request) {
        Owner owner = owner(context);
        if (repository.countByTenantIdAndOwnerUsername(owner.tenantId(), owner.username()) >= MAX_DOCUMENTS_PER_USER) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "云端流程图数量已达到 100 个上限");
        }
        byte[] snapshot = decodeSnapshot(request.update());
        String now = Instant.now().toString();
        UserDiagramDocument document = new UserDiagramDocument();
        document.setId(newUniqueId());
        document.setTenantId(owner.tenantId());
        document.setOwnerUsername(owner.username());
        document.setName(requireText(request.name(), "name", 120));
        document.setSnapshotData(snapshot);
        document.setSizeBytes(snapshot.length);
        document.setCreatedAt(now);
        document.setUpdatedAt(now);
        return view(repository.saveAndFlush(document));
    }

    @Transactional
    public DiagramDocumentView update(ManagementContext context, long id, DiagramDocumentMutation request) {
        UserDiagramDocument document = requireOwned(context, id);
        if (request.revision() == null || request.revision() != document.getRevision()) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "云端文件已被其他会话更新，请重新打开后再保存");
        }
        byte[] snapshot = decodeSnapshot(request.update());
        document.setName(requireText(request.name(), "name", 120));
        document.setSnapshotData(snapshot);
        document.setSizeBytes(snapshot.length);
        document.setUpdatedAt(Instant.now().toString());
        try {
            return view(repository.saveAndFlush(document));
        } catch (ObjectOptimisticLockingFailureException conflict) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "云端文件已被其他会话更新，请重新打开后再保存", conflict);
        }
    }

    @Transactional
    public void delete(ManagementContext context, long id) {
        repository.delete(requireOwned(context, id));
    }

    private UserDiagramDocument requireOwned(ManagementContext context, long id) {
        Owner owner = owner(context);
        return repository.findByIdAndTenantIdAndOwnerUsername(id, owner.tenantId(), owner.username())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "云端流程图不存在"));
    }

    private Owner owner(ManagementContext context) {
        if (context == null || context.tenant() == null) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "需要登录后才能访问云端流程图");
        }
        return new Owner(
                requireText(context.tenant().tenantId(), "tenantId", 80),
                requireText(context.username(), "username", 160)
        );
    }

    private byte[] decodeSnapshot(String encoded) {
        if (!StringUtils.hasText(encoded) || encoded.length() > 4 * 1024 * 1024 + 16) {
            throw new IllegalArgumentException("流程图数据无效或超过限制");
        }
        byte[] decoded;
        try {
            decoded = Base64.getDecoder().decode(encoded);
        } catch (IllegalArgumentException exception) {
            throw new IllegalArgumentException("流程图数据不是有效的 Base64", exception);
        }
        if (decoded.length == 0 || decoded.length > MAX_SNAPSHOT_BYTES) {
            throw new IllegalArgumentException("流程图数据无效或超过 3 MB");
        }
        return decoded;
    }

    private long newUniqueId() {
        for (int attempt = 0; attempt < 8; attempt++) {
            long id = ClientIdGenerator.newId();
            if (!repository.existsById(id)) return id;
        }
        throw new IllegalStateException("无法生成云端流程图 ID");
    }

    private DiagramDocumentView view(UserDiagramDocument document) {
        return new DiagramDocumentView(
                document.getId(),
                document.getName(),
                document.getSizeBytes(),
                document.getRevision(),
                document.getCreatedAt(),
                document.getUpdatedAt()
        );
    }

    private DiagramDocumentView view(UserDiagramDocumentSummary document) {
        return new DiagramDocumentView(
                document.getId(),
                document.getName(),
                document.getSizeBytes(),
                document.getRevision(),
                document.getCreatedAt(),
                document.getUpdatedAt()
        );
    }

    private static String requireText(String value, String field, int maxLength) {
        if (!StringUtils.hasText(value)) throw new IllegalArgumentException(field + " 不能为空");
        String normalized = value.trim();
        if (normalized.length() > maxLength) throw new IllegalArgumentException("字段长度不能超过 " + maxLength);
        if (normalized.indexOf('\r') >= 0 || normalized.indexOf('\n') >= 0) {
            throw new IllegalArgumentException("字段不能包含换行");
        }
        return normalized;
    }

    private record Owner(String tenantId, String username) {}

    public record DiagramDocumentMutation(String name, String update, Long revision) {}

    public record DiagramDocumentView(long id,
                                      String name,
                                      long sizeBytes,
                                      long revision,
                                      String createdAt,
                                      String updatedAt) {}

    public record DiagramDocumentDetail(DiagramDocumentView document, String update) {}
}

File: implementations/csharp/server/src/Specus.Server/Management/UserDiagramDocumentService.cs
using Microsoft.EntityFrameworkCore;
using Specus.Server.Authentication;
using Specus.Server.Data;
using Specus.Server.Data.Entities;

namespace Specus.Server.Management;

public sealed class ResourceNotFoundException(string message) : Exception(message);

public sealed class UserDiagramDocumentService
{
    private const int MaxDocumentsPerUser = 100;
    private const int MaxSnapshotBytes = 3 * 1024 * 1024;

    private readonly SpecusDbContext _db;

    public UserDiagramDocumentService(SpecusDbContext db)
    {
        _db = db;
    }

    public async Task<IReadOnlyList<DiagramDocumentView>> ListAsync(ManagementContext context,
        CancellationToken cancellationToken)
    {
        var owner = Owner(context);
        var rows = await _db.UserDiagramDocuments.AsNoTracking()
            .Where(row => row.TenantId == owner.TenantId && row.OwnerUsername == owner.Username)
            .OrderByDescending(row => row.UpdatedAt)
            .ToListAsync(cancellationToken)
            .ConfigureAwait(false);
        return rows.Select(View).ToList();
    }

    public async Task<DiagramDocumentDetail> GetAsync(ManagementContext context, long id,
        CancellationToken cancellationToken)
    {
        var document = await RequireOwnedAsync(context, id, tracking: false, cancellationToken)
            .ConfigureAwait(false);
        return new DiagramDocumentDetail(View(document), Convert.ToBase64String(document.SnapshotData));
    }

    public async Task<DiagramDocumentView> CreateAsync(ManagementContext context,
        DiagramDocumentMutation request, CancellationToken cancellationToken)
    {
        var owner = Owner(context);
        if (await _db.UserDiagramDocuments.AsNoTracking()
                .LongCountAsync(row => row.TenantId == owner.TenantId
                    && row.OwnerUsername == owner.Username, cancellationToken)
                .ConfigureAwait(false) >= MaxDocumentsPerUser)
        {
            throw new InvalidOperationException("云端流程图数量已达到 100 个上限");
        }

        var snapshot = DecodeSnapshot(request.Update);
        var now = DateTimeOffset.UtcNow;
        var document = new UserDiagramDocument
        {
            Id = await NewUniqueIdAsync(cancellationToken).ConfigureAwait(false),
            TenantId = owner.TenantId,
            OwnerUsername = owner.Username,
            Name = RequireText(request.Name, "name", 120),
            SnapshotData = snapshot,
            SizeBytes = snapshot.LongLength,
            Revision = 0,
            CreatedAt = now,
            UpdatedAt = now,
        };
        _db.UserDiagramDocuments.Add(document);
        await _db.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
        return View(document);
    }

    public async Task<DiagramDocumentView> UpdateAsync(ManagementContext context, long id,
        DiagramDocumentMutation request, CancellationToken cancellationToken)
    {
        var document = await RequireOwnedAsync(context, id, tracking: true, cancellationToken)
            .ConfigureAwait(false);
        if (request.Revision is null || request.Revision.Value != document.Revision)
        {
            throw Conflict();
        }

        var snapshot = DecodeSnapshot(request.Update);
        document.Name = RequireText(request.Name, "name", 120);
        document.SnapshotData = snapshot;
        document.SizeBytes = snapshot.LongLength;
        document.Revision++;
        document.UpdatedAt = DateTimeOffset.UtcNow;
        try
        {
            await _db.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (DbUpdateConcurrencyException conflict)
        {
            throw Conflict(conflict);
        }
        return View(document);
    }

    public async Task DeleteAsync(ManagementContext context, long id,
        CancellationToken cancellationToken)
    {
        var document = await RequireOwnedAsync(context, id, tracking: true, cancellationToken)
            .ConfigureAwait(false);
        _db.UserDiagramDocuments.Remove(document);
        await _db.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
    }

    private async Task<UserDiagramDocument> RequireOwnedAsync(ManagementContext context, long id,
        bool tracking, CancellationToken cancellationToken)
    {
        var owner = Owner(context);
        var query = _db.UserDiagramDocuments
            .Where(row => row.Id == id && row.TenantId == owner.TenantId
                && row.OwnerUsername == owner.Username);
        if (!tracking)
        {
            query = query.AsNoTracking();
        }
        return await query.FirstOrDefaultAsync(cancellationToken).ConfigureAwait(false)
            ?? throw new ResourceNotFoundException("云端流程图不存在");
    }

    private async Task<long> NewUniqueIdAsync(CancellationToken cancellationToken)
    {
        for (var attempt = 0; attempt < 8; attempt++)
        {
            var id = ClientIdGenerator.NewId();
            if (!await _db.UserDiagramDocuments.AsNoTracking()
                    .AnyAsync(row => row.Id == id, cancellationToken).ConfigureAwait(false))
            {
                return id;
            }
        }
        throw new InvalidOperationException("无法生成云端流程图 ID");
    }

    private static byte[] DecodeSnapshot(string? encoded)
    {
        if (string.IsNullOrWhiteSpace(encoded) || encoded.Length > 4 * 1024 * 1024 + 16)
        {
            throw new ArgumentException("流程图数据无效或超过限制");
        }
        byte[] decoded;
        try
        {
            decoded = Convert.FromBase64String(encoded);
        }
        catch (FormatException exception)
        {
            throw new ArgumentException("流程图数据不是有效的 Base64", exception);
        }
        if (decoded.Length == 0 || decoded.Length > MaxSnapshotBytes)
        {
            throw new ArgumentException("流程图数据无效或超过 3 MB");
        }
        return decoded;
    }

    private static (string TenantId, string Username) Owner(ManagementContext context) =>
        (RequireText(context.TenantId, "tenantId", 80),
            RequireText(context.Username, "username", 160));

    private static string RequireText(string? value, string field, int maxLength)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new ArgumentException($"{field} 不能为空");
        }
        var normalized = value.Trim();
        if (normalized.Length > maxLength)
        {
            throw new ArgumentException($"字段长度不能超过 {maxLength}");
        }
        if (normalized.Contains('\r') || normalized.Contains('\n'))
        {
            throw new ArgumentException("字段不能包含换行");
        }
        return normalized;
    }

    private static InvalidOperationException Conflict(Exception? inner = null) =>
        new("云端文件已被其他会话更新，请重新打开后再保存", inner);

    private static DiagramDocumentView View(UserDiagramDocument document) => new(
        document.Id, document.Name, document.SizeBytes, document.Revision,
        Iso(document.CreatedAt), Iso(document.UpdatedAt));

    private static string Iso(DateTimeOffset value) => value.UtcDateTime.ToString("O");
}

public sealed record DiagramDocumentMutation(string? Name, string? Update, long? Revision);

public sealed record DiagramDocumentView(
    long Id,
    string Name,
    long SizeBytes,
    long Revision,
    string CreatedAt,
    string UpdatedAt);

public sealed record DiagramDocumentDetail(DiagramDocumentView Document, string Update);
